### PR TITLE
feat: onSwapQuote handler

### PR DESCRIPTION
**Pull request by zzmp**
Adds a handler for swap quote timings, to be used by consumers for performance metrics.

### CODE REVIEW

File: src/state/routing/slice.ts [hunk 1]
```diff
@@ -23,9 +23,32 @@ const baseQuery: BaseQueryFn<GetQuoteArgs, TradeResult> = () => {
...
+        } finally {
+          const duration = performance.now() - timestamp
+          args.onQuote?.(
+            {
+              timestamp,
+              duration,
+              args: JSON.parse(serializeGetQuoteArgs(args)),
+              state: result?.state,
+            },
+            error
```
Comment (zzmp) [inline src/state/routing/slice.ts:48]: This can be narrowed to `FetchBaseQueryError`

File: src/state/routing/args.ts [hunk 2]
```diff
@@ -6,17 +6,15 @@ import { QuoteConfig, QuoteType } from 'hooks/routing/types'
...
 import { currencyAddressForSwapQuote } from './utils'
 
-const NON_SERIALIZABLE_KEYS = ['provider']
+const NON_SERIALIZABLE_KEYS = ['provider', 'onQuote']
 
-/** Omits the non-serializable keys from GetQuoteArgs' cache key. */
-export function serializeGetQuoteArgs({ endpointName, queryArgs }: { endpointName: string; queryArgs: GetQuoteArgs }) {
-  // same as default serializeQueryArgs, but ignoring non-serializable keys.
-  return `${endpointName}(${JSON.stringify(queryArgs, (key, value) => {
+export function serializeGetQuoteArgs(args: GetQuoteArgs) {
```
Comment (just-toby) [inline src/state/routing/args.ts:16]: is this function specific to `onQuote`/`GetQuote` or can we just rename it to be more generic (might help disambiguate between it and the new function below)

File: src/state/swap/index.ts [hunk 3]
```diff
@@ -85,6 +85,17 @@ interface InputEventHandlers {
 /** An integration hook called when the user approves a token, either through allowance or permit. */
 export type OnSwapApprove = () => void
 
+/** An integration hook called when a new quote is fetched. */
+export type OnSwapQuote = (
+  event: {
+    timestamp: number
```
Comment (just-toby) [inline src/state/swap/index.ts:91]: is this the timestamp of the query start or end? (unclear to integrator)

File: src/state/swap/index.ts [hunk 4]
```diff
@@ -85,6 +85,17 @@ interface InputEventHandlers {
 /** An integration hook called when the user approves a token, either through allowance or permit. */
 export type OnSwapApprove = () => void
 
+/** An integration hook called when a new quote is fetched. */
+export type OnSwapQuote = (
```
Comment (just-toby) [inline src/state/swap/index.ts:89]: do you think there's any benefit to making `event` fields optional? i'm thinking of situations where we need to change the API (adding/removing fields without disrupting integrators). maybe this only applies to when we add or remove later?

**Review by just-toby — APPROVED** [review]
lgtm

File: src/state/routing/args.ts (hunk 2 above)
Comment (zzmp) [reply src/state/routing/args.ts:16]: These are specific to `GetQuote`:
- `serializeGetQuoteArgs` serializes a `GetQuoteArgs` object
- `serializeGetQuoteQueryArgs` is plugged into `serializeQueryArgs` for the `GetQuote` query

I think they're both named appropriately, just unfortunately very similarly.

File: src/state/swap/index.ts (hunk 4 above)
Comment (zzmp) [reply src/state/swap/index.ts:89]: I don't think there's a benefit - I don't think there's a large issue with changing this down the road.

**Pull request closed without merging**